src/jdk.internal.md/share/classes/jdk/internal/markdown/MarkdownTransformer.java

```diff
@@ -0,0 +1,951 @@
+                var item1 = iter1.next();
+                var item2 = iter2.next();
+                if (item1 != item2) {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+    }
+
+    private static final String AUTOREF_PREFIX = "code:";
+
+    /**
+     * An {@code InlineParserFactory} that checks any unresolved link references in
+     * reference links. If an unresolved link reference appears to be a reference to a
+     * program element, such as may be used in {@code @see ...} or {@code {@link ...}}
+     * tags, it generates a synthetic {@code LinkReferenceDefinition}.
+     *
+     * The reference is not validated to ensure it refers to any existing element.
+     *
+     * The primary function of this {@code InlineParserFactory} is to provide
+     * a custom implementation of an {@code InlineParserContext} via the
+     * {@code create(InlineParserContext inlineParserContext)} method.
+     * It is this {@code InlineParserContext} that actually manages the set
+     * of link reference definitions.
+     *
+     * This {@code InlineParserFactory} is intended to be registered with a
+     * {@code Parser.Builder} using {@code setInlineParserFactory}.
+     */
+    private static class AutoRefInlineParserFactory implements InlineParserFactory {
+        private final ReferenceParser refParser;
+
+        AutoRefInlineParserFactory(ReferenceParser refParser) {
+            this.refParser = refParser;
+        }
+
+        /**
+         * Creates a parser with a modified {@code InlineParserContext} that
+         * delegates to the standard {@code InlineParserContext} and also
+         * checks unresolved link references.
+         *
+         * @param inlineParserContext the standard {@code InlineParserContext}
+         * @return the {@code InlineParser}
+         */
+        @Override
+        public InlineParser create(InlineParserContext inlineParserContext) {
+            return new InlineParserImpl(new InlineParserContext() {
+
+                @Override
+                public List<DelimiterProcessor> getCustomDelimiterProcessors() {
+                    return inlineParserContext.getCustomDelimiterProcessors();
+                }
+
+                /**
+                 * {@inheritDoc}
+                 *
+                 * If the given label does not match any explicitly defined
+                 * link reference definition, but does match a reference
+                 * to a program element, a synthetic link reference definition
+                 * is returned, with the label prefixed by {@code code:}.
+                 *
+                 * @param label the link label to look up
+                 * @return the link reference definition for the label, or {@code null}
+                 */
+                @Override
+                public LinkReferenceDefinition getLinkReferenceDefinition(String label) {
+                    var l = label.replace("\\[\\]", "[]");
```

Comment:
Yes, you need to escape `[` and `]` within the label of any Markdown reference link, by preceding each with backslash.   (Remember, the label is the string used to find the URL for the link; not the displayed text of the link).
That's a CommonMark feature independent of this work.

While we could change that `replace` call into two separate ones, in reference signatures they always appear together as a pair, and can be replaced together.   We need to remove the escape characters in the generated URL so that the signature is a standard signature with unescaped `[]`

For fun/demo, try the following:

```
/// First sentence.
/// * Link 0 to [java.lang.Object]
/// * Link 1a to [Arrays-equals][java.util.Arrays#equals(Object[],Object[])]
/// * Link 1b to [Arrays-equals][java.util.Arrays#equals(Object\[\],Object\[\])]
/// * Link 2a to [java.util.Arrays#equals(Object[],Object[])]
/// * Link 2b to [java.util.Arrays#equals(Object\[\],Object\[\])]
public class C { }
```

Link 1a and 2a end up as unprocessed "literal" text (because the `[]` were not escaped.)  That is, they are not even recognized by the CommonMark parser as reference links.  Link 1b and 2b get processed as links, as expected.

FWIW, this issue with needing to escape `[]` pairs is specifically mentioned in the JEP as an inescapable (sic) limitation.

Reply:
I'll add a test case.

Reply:
Done

Reply:
Thank you. I double-checked: if that `replace` is removed, jdk/javadoc/doclet/testMarkdown/TestMarkdown.java fails.